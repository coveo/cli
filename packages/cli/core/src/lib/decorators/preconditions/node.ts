import {getBinVersionPrecondition} from './binPreconditionsFactory.js';

export const IsNodeVersionInRange = getBinVersionPrecondition('node', {
  prettyName: 'Node.js',
});
