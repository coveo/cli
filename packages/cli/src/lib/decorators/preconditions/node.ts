import {constants} from 'os';
import * as dedent from 'dedent';

import type Command from '@oclif/command';
import {valid as isVersionValid, lt as isVersionLessThan} from 'semver';

import {SpawnProcessOutput, spawnProcessOutput} from '../../utils/process';

const linkToReadme =
  'https://github.com/coveo/cli/wiki/Node.js,-NPM-and-NPX-requirements';

export function IsNodeVersionAbove(requiredVersion: string) {
  return async function (target: Command) {
    if (!isVersionValid(requiredVersion)) {
      target.warn(dedent`
        Required version invalid: "${requiredVersion}".
        Please report this error to Coveo: https://github.com/coveo/cli/issues/new
      `);
      return false;
    }

    const output = await spawnProcessOutput('node', ['--version']);

    if (isMissingExecutable(output)) {
      target.warn(`${target.id} requires Node.js to run.`);
      warnHowToInstallNode(target);
      return false;
    }

    if (output.stderr || output.exitCode !== 0) {
      target.warn(dedent`
        ${target.id} requires a valid Node.js installation to run.
        An unknown error happened while trying to determine your node version with node --version
        ${output.stderr}
      `);
      warnHowToInstallNode(target);
      return false;
    }

    if (isVersionLessThan(output.stdout, requiredVersion)) {
      target.warn(dedent`
        ${target.id} needs a Node.js version greater than ${requiredVersion}
        Version detected: ${output.stdout}
      `);
      warnHowToInstallNode(target);
      return false;
    }

    return true;
  };
}

export function warnHowToInstallNode(target: Command) {
  target.warn(dedent`
    Please visit ${linkToReadme} for more detailed installation information.
  `);
}

export function isMissingExecutable(output: SpawnProcessOutput) {
  return (
    output.exitCode !== null &&
    Math.abs(output.exitCode) === constants.errno.ENOENT
  );
}
