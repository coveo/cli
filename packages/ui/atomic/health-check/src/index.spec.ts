import {spawnSync} from 'node:child_process';
import {cpSync} from 'fs-extra';
import {mkdirSync, emptyDir} from 'fs-extra';
import {join} from 'node:path';
import {DirResult, dirSync} from 'tmp';

const execPath = join(__dirname, '..', 'dist');

const pathToStub = join(__dirname, '..', '__stub__');
const healthCheck = (cwd: string) =>
  spawnSync('node', [execPath], {
    cwd,
    shell: process.platform === 'win32' ? 'powershell' : undefined,
  });

const addPackageJsonToProject = (stubFile: string, dest: string) => {
  cpSync(join(pathToStub, stubFile), join(dest, 'package.json'));
};

const addReadmeToProject = (dest: string) => {
  cpSync(join(pathToStub, 'readme.md'), join(dest, 'readme.md'));
};

const addJsonDocsToProject = (stubFile: string, dest: string) => {
  cpSync(join(pathToStub, stubFile), join(dest, 'docs', 'stencil-docs.json'));
};

describe('@coveo/atomic-component-health-check', () => {
  let tempDirectory: DirResult;
  let testDirectory: string;

  beforeAll(async () => {
    tempDirectory = dirSync({unsafeCleanup: true, keep: true});
    testDirectory = join(tempDirectory.name, 'dummy-project');
    mkdirSync(testDirectory, {recursive: true});
  });

  beforeEach(() => {
    emptyDir(testDirectory);
  });

  afterAll(async () => {
    tempDirectory.removeCallback();
  });

  describe('when readme condition is not met', () => {
    beforeAll(() => {
      addJsonDocsToProject('matching-docs.json', testDirectory);
      addPackageJsonToProject('validPackage.json', testDirectory);
    });

    it('should exit with status code 1', () => {
      const {status} = healthCheck(testDirectory);
      expect(status).toBe(1);
    });

    it('should warn about missing readme file', () => {
      const {stdout} = healthCheck(testDirectory);
      expect(stdout.toString()).toMatchSnapshot();
    });
  });

  describe('when package.json conditions are not met', () => {
    beforeEach(() => {
      addReadmeToProject(testDirectory);
    });

    it('should exit with status code 1', () => {
      addPackageJsonToProject('missingProperties.json', testDirectory);
      const {status} = healthCheck(testDirectory);
      expect(status).toBe(1);
    });

    it.each([
      {
        title: 'should warn about missing properties',
        stubPackage: 'missingProperties.json',
      },
      {
        title: 'should warn about invalid properties',
        stubPackage: 'invalidProperties.json',
      },
    ])('$title', ({stubPackage}) => {
      addPackageJsonToProject(stubPackage, testDirectory);
      const {stdout} = healthCheck(testDirectory);
      expect(stdout.toString()).toMatchSnapshot();
    });
  });

  describe('when component tag name does not match `elementName` property', () => {
    beforeAll(() => {
      addJsonDocsToProject('clashing-docs.json', testDirectory);
      addPackageJsonToProject('validPackage.json', testDirectory);
    });

    it('should exit with status code 1', () => {
      const {status} = healthCheck(testDirectory);
      expect(status).toBe(1);
    });

    it('should warn about component name mismatch', () => {
      const {stdout} = healthCheck(testDirectory);
      expect(stdout.toString()).toMatchSnapshot();
    });
  });

  describe('when all conditions are met', () => {
    beforeEach(() => {
      addJsonDocsToProject('matching-docs.json', testDirectory);
      addReadmeToProject(testDirectory);
      addPackageJsonToProject('validPackage.json', testDirectory);
    });

    it('should print green checks only', () => {
      const {stdout} = healthCheck(testDirectory);
      expect(stdout.toString()).toMatchSnapshot();
    });
  });
});
