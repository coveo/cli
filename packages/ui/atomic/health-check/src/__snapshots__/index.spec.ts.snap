// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@coveo/atomic-component-health-check when all conditions are met should print green checks only 1`] = `
"[32mâœ”[39m [1mReadme file[22m

[32mâœ”[39m [1mComponent name[22m

[32mâœ”[39m [1mRequired properties in package.json[22m

"
`;

exports[`@coveo/atomic-component-health-check when component tag name does not match \`elementName\` property should warn about component name mismatch 1`] = `
"[32mâœ”[39m [1mReadme file[22m

[31mâœ–[39m [1mComponent name[22m
  Component tag name from your .tsx file does not match the \`elementName\` property defined in your component's package.json file. Make sure both values are identical

[32mâœ”[39m [1mRequired properties in package.json[22m

Publish aborted because some conditions have not been met
Make sure to address the above errors before publishing again
"
`;

exports[`@coveo/atomic-component-health-check when package.json conditions are not met should warn about invalid properties 1`] = `
"[32mâœ”[39m [1mReadme file[22m

[31mâœ–[39m [1mComponent name[22m
  Component tag name from your .tsx file does not match the \`elementName\` property defined in your component's package.json file. Make sure both values are identical

[31mâœ–[39m [1mRequired properties in package.json[22m
    [31mâœ–[39m [1mInvalid elementName: [22mYou must provide a \`elementName\` property for your custom component. Its value should match the tag name used by your custom component. See https://stenciljs.com/docs/component#tag
    [31mâœ–[39m [1mInvalid keywords: [22mThe package.json \`keywords\` array should include \`coveo-atomic-component\`, otherwise your component will not be listed along with other Coveo custom components. See https://docs.npmjs.com/cli/v9/configuring-npm/package-json#keywords
    [31mâœ–[39m [1mInvalid description: [22mThe component description must be at least 20 characters long

Publish aborted because some conditions have not been met
Make sure to address the above errors before publishing again
"
`;

exports[`@coveo/atomic-component-health-check when package.json conditions are not met should warn about missing properties 1`] = `
"[32mâœ”[39m [1mReadme file[22m

[31mâœ–[39m [1mComponent name[22m
  Component tag name from your .tsx file does not match the \`elementName\` property defined in your component's package.json file. Make sure both values are identical

[31mâœ–[39m [1mRequired properties in package.json[22m
    [31mâœ–[39m [1mInvalid elementName: [22mYou must provide a \`elementName\` property for your custom component. Its value should match the tag name used by your custom component. See https://stenciljs.com/docs/component#tag
    [31mâœ–[39m [1mInvalid unpkg: [22mYou must provide the path to the \`.esm.js\` file of your custom component. You can find it under the \`dist/\` folder once you build the component.
    [31mâœ–[39m [1mInvalid keywords: [22mYou must populate the \`keywords\` property. See https://docs.npmjs.com/cli/v9/configuring-npm/package-json#keywords
    [31mâœ–[39m [1mInvalid description: [22mYou must provide a description at least 20 characters long to explain the component purpose.
    [31mâœ–[39m [1mInvalid homepage: [22mYou must provide a URL to the component source code. See https://docs.npmjs.com/cli/v9/configuring-npm/package-json#homepage

Publish aborted because some conditions have not been met
Make sure to address the above errors before publishing again
"
`;

exports[`@coveo/atomic-component-health-check when readme condition is not met should warn about missing readme file 1`] = `
"[31mâœ–[39m [1mReadme file[22m
  Missing README file. Make sure to include a \`README.md\` file in your component directory

[32mâœ”[39m [1mComponent name[22m

[32mâœ”[39m [1mRequired properties in package.json[22m

Publish aborted because some conditions have not been met
Make sure to address the above errors before publishing again
"
`;
